EIPS/eip-3157.md

```diff
@@ -0,0 +1,265 @@
+The interfaces described in this ERC have been chosen as to cover the known flash lending use cases, while allowing for safe and gas efficient implementations.
+
+`flashFee` reverts on unsupported tokens, because returning a numerical value would be incorrect.
+
+`flashLoan` has been chosen as a function name as descriptive enough, unlikely to clash with other functions in the lender, and including both the use cases in which the tokens lended are held or minted by the lender.
+
+`receiver` is taken as a parameter to allow flexibility on the implementation of separate loan initiators and receivers.
+
+Existing flash lenders (Aave, dYdX and Uniswap) all provide flash loans of several token types from the same contract (LendingPool, SoloMargin and UniswapV2Pair). Providing a `token` parameter in both the `flashLoan` and `onFlashLoan` functions matches closely the observed functionality.
+
+A `bytes calldata data` parameter is included for the caller to pass arbitrary information to the `receiver`, without impacting the utility of the `flashLoan` standard.
+
+`onFlashLoan` has been chosen as a function name as descriptive enough, unlikely to clash with other functions in the `receiver`, and following the `onAction` naming pattern used as well in EIP-667.
+
+A `sender` will often be required in the `onFlashLoan` function, which the lender knows as `msg.sender`. An alternative implementation which would embed the `sender` in the `data` parameter by the caller would require an additional mechanism for the receiver to verify its accuracy, and is not advisable.
+
+The `amount` will be required in the `onFlashLoan` function, which the lender took as a parameter. An alternative implementation which would embed the `amount` in the `data` parameter by the caller would require an additional mechanism for the receiver to verify its accuracy, and is not advisable.
+
+A `fee` will often be calculated in the `flashLoan` function, which the `receiver` must be aware of for repayment. Passing the `fee` as a parameter instead of appended to `data` is simple and effective.
+
+The `amount + fee` are pulled from the `receiver` to allow the `lender` to implement other features that depend on using `transferFrom`, without having to lock them for the duration of a flash loan. An alternative implementation where the repayment is transferred to the `lender` is also possible, but would need all other features in the lender to be also based in using `transfer` instead of `transferFrom`. Given the lower complexity and prevalence of a "pull" architecture over a "push" architecture, "pull" was chosen.
+
+## Backwards Compatibility
+
+No backwards compatibility issues identified.
+
+## Implementation
+
+### Flash Borrower Reference Implementation
+
+TBA
+
+### Flash Mint Reference Implementation
+
+TBA - Flash mints make less sense as part of a batch flash loan
+
+### Flash Loan Reference Implementation
+
+TBA
+
+## Security Considerations
+
+_Copy from ERC3156, or reference it?_
```

Comment:
Recommend copying.  Each EIP should stand on its own as much as possible unless it has a hard dependency.

Reply:
Copied.